### PR TITLE
Add AMP validation features to support literal attribute value, e.g [].

### PR DESCRIPTION
**Pull request by nhant01**
1. When inspecting the attributes for p, the "[text]" attribute is returned as "text", when these are distinct values. We are requesting a change or some feature that can return the literal value of the attribute. This is necessary for AMP validation.

2. At the moment, tagchowder is sanitizing Unicode characters by replacing with "_". For example, a tag that looks like,
'html ⚡4email' would be returned at the handler level as, 'html _4email'

### CODE REVIEW

File: core/src/main/java/com/yahoo/tagchowder/Parser.java [hunk 1]
```diff
@@ -166,6 +168,11 @@
      **/
     public static final String VALIDATION_FEATURE = "http://xml.org/sax/features/validation";
 
+    /**
+     * Controls whether the parser is reporting all validity errors for AMP contents.
+     */
+    public static final String AMP_VALIDATION_FEATURE = "http://xml.org/sax/features/amp-validation";
```
Comment (squintview) [inline core/src/main/java/com/yahoo/tagchowder/Parser.java:174]: if this is just a dummy string, we might as well put the amp java validator git link

File: core/src/main/java/com/yahoo/tagchowder/Parser.java [hunk 2]
```diff
@@ -1185,6 +1195,9 @@ private String makeName(final char[] buff, final int offset, final int length) {
                 }
                 start = true;
                 dst.append(translateColons ? '_' : ch);
+            } else if (ampValidation && (ch == '[' || ch == ']' || ch == '{' || ch == '}' || ch == '⚡')) {
```
Comment (squintview) [inline core/src/main/java/com/yahoo/tagchowder/Parser.java:1198]: Shouldn't we allow [, { and ⚡ only if "start=true"?

Comment (squintview) [reply core/src/main/java/com/yahoo/tagchowder/Parser.java:1198]: ], } can be allowed if its the last char in the attribute name

File: core/src/main/java/com/yahoo/tagchowder/Parser.java (hunk 1 above)
Comment (nhant01) [reply core/src/main/java/com/yahoo/tagchowder/Parser.java]: Done

File: core/src/main/java/com/yahoo/tagchowder/Parser.java (hunk 2 above)
Comment (nhant01) [reply core/src/main/java/com/yahoo/tagchowder/Parser.java]: Good point! fixed

File: core/src/main/java/com/yahoo/tagchowder/Parser.java [hunk 3]
```diff
@@ -280,6 +287,7 @@
         theFeatures.put(RESTART_ELEMENTS_FEATURE, truthValue(DEFAULT_RESTART_ELEMENTS));
         theFeatures.put(IGNORABLE_WHITESPACE_FEATURE, truthValue(DEFAULT_IGNORABLE_WHITESPACE));
         theFeatures.put(CDATA_ELEMENTS_FEATURE, truthValue(DEFAULT_CDATA_ELEMENTS));
+        theFeatures.put(AMP_VALIDATION_FEATURE, truthValue(DEFAULT_AMP_VALIDATION));
```
Comment (fansu) [inline core/src/main/java/com/yahoo/tagchowder/Parser.java:290]: add one UT for amp feature
1) use getFeature() funtion to check the default value of AMP_VALIDATION_FEATURE is false
2) call setFeature() to set AMP_VALIDATION_FEATURE, call getFeature() again, it should return true

File: core/src/main/java/com/yahoo/tagchowder/Parser.java [hunk 4]
```diff
@@ -1185,6 +1195,14 @@ private String makeName(final char[] buff, final int offset, final int length) {
                 }
                 start = true;
                 dst.append(translateColons ? '_' : ch);
+            } else if (ampValidation) {
+                if (start && (ch == '[' || ch == '{' || ch == '⚡'))  {
```
Comment (fansu) [inline core/src/main/java/com/yahoo/tagchowder/Parser.java:1199]: based on your code, you are going to accept and not sanitize following html? These examples are not full cases. Please test it fully and have document expected behavior and add corresponding UTs.

<p [text}="cart.p1_qty" class="qty_num">1</p>
<p {text}="cart.p1_qty" class="qty_num">1</p>
<p {text]="cart.p1_qty" class="qty_num">1</p>
<p {text="cart.p1_qty" class="qty_num">1</p>
<p text}="cart.p1_qty" class="qty_num">1</p>
<p [text="cart.p1_qty" class="qty_num">1</p>
<p text]="cart.p1_qty" class="qty_num">1</p>
<p text]="cart.p1_qty" class="qty_num">1</p>
<p ⚡text]="cart.p1_qty" class="qty_num">1</p>
<p ⚡text}="cart.p1_qty" class="qty_num">1</p>
<p ⚡text="cart.p1_qty" class="qty_num">1</p>
<p ⚡text="cart.p1_qty" class="qty_num">1</p>

**Review by fansu — CHANGES_REQUESTED** [review]
Please add test to well cover the new code

File: core/src/main/java/com/yahoo/tagchowder/Parser.java (hunk 3 above)
Comment (nhant01) [reply core/src/main/java/com/yahoo/tagchowder/Parser.java:297]: Added

File: core/src/main/java/com/yahoo/tagchowder/Parser.java (hunk 4 above)
Comment (nhant01) [reply core/src/main/java/com/yahoo/tagchowder/Parser.java]: Redo the implementation to ensure brackets are balanced and check for "⚡4email" validity.

File: core/src/main/java/com/yahoo/tagchowder/Parser.java [hunk 5]
```diff
@@ -1185,6 +1204,26 @@ private String makeName(final char[] buff, final int offset, final int length) {
...
+                //if the beginning character starts with '[' or '{', ensure the brackets are balanced.
+                //allows [text], [[text]], {text}, {{text}}..etc
+                if (start && (ch == '[' || ch == '{'))  {
+                    if (areBracketsBalanced(buff, offset, length)) {
+                        dst.append(ch);
+                        bracketsAreBalanced = true;
+                    }
+                } else if (start && ch == '⚡') {
+                    //if the beginning character starts with '⚡', ensure that this buffer with offset and length contains "⚡4email"
+                    final StringBuffer str = new StringBuffer();
```
Comment (fansu) [inline core/src/main/java/com/yahoo/tagchowder/Parser.java:1217]: please allocate length for stringbuffer

Comment (nhant01) [reply core/src/main/java/com/yahoo/tagchowder/Parser.java]: Done

File: core/src/main/java/com/yahoo/tagchowder/Parser.java [hunk 6]
```diff
@@ -1185,6 +1204,26 @@ private String makeName(final char[] buff, final int offset, final int length) {
                 }
                 start = true;
                 dst.append(translateColons ? '_' : ch);
+            } else if (ampValidation) {
+                //if the beginning character starts with '[' or '{', ensure the brackets are balanced.
+                //allows [text], [[text]], {text}, {{text}}..etc
+                if (start && (ch == '[' || ch == '{'))  {
+                    if (areBracketsBalanced(buff, offset, length)) {
```
Comment (fansu) [inline core/src/main/java/com/yahoo/tagchowder/Parser.java:1211]: The more I read this code, the more I feel it is not that efficient. Looks like you can going to scan the whole string whenever there is a '[' multiple times in following case. Is there any way to make it more efficient? 
[[[[[[[[[[[text]]]]]]]]]]
Maybe you can scan it once for areBracketsBalanced() and use that info for real scanning here?

File: core/src/main/java/com/yahoo/tagchowder/Parser.java [hunk 7]
```diff
@@ -1185,6 +1204,26 @@ private String makeName(final char[] buff, final int offset, final int length) {
...
+                        dst.append(ch);
+                        bracketsAreBalanced = true;
+                    }
+                } else if (start && ch == '⚡') {
+                    //if the beginning character starts with '⚡', ensure that this buffer with offset and length contains "⚡4email"
+                    final StringBuffer str = new StringBuffer(length);
+                    str.append(buff, offset, length);
+                    if (AMP4EMAIL.equalsIgnoreCase(str.toString())) {
+                        dst.append(ch);
```
Comment (fansu) [inline core/src/main/java/com/yahoo/tagchowder/Parser.java:1220]: the whole string has been compared,right? you do not need to continue, you can break

File: core/src/main/java/com/yahoo/tagchowder/Parser.java (hunk 6 above)
Comment (nhant01) [reply core/src/main/java/com/yahoo/tagchowder/Parser.java:1211]: There should be one scan. The start flag changes to false.

File: core/src/main/java/com/yahoo/tagchowder/Parser.java (hunk 7 above)
Comment (nhant01) [reply core/src/main/java/com/yahoo/tagchowder/Parser.java]: Done

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java [hunk 8]
```diff
@@ -51,6 +52,69 @@ public void testSampleHtml() throws IOException, SAXException {
...
+     * @throws IOException IOException
+     * @throws SAXException SAXException
+     */
+    @Test
+    public void testSampleAMPHtml() throws IOException, SAXException {
+        final String html = getSampleHtml("amphtml.txt");
+        final Parser parser = new Parser();
+        parser.setFeature(Parser.AMP_VALIDATION_FEATURE, true);
+        final InputSource inSource = new InputSource(new StringReader(html));
+        parser.parse(inSource);
```
Comment (fansu) [inline core/src/test/java/com/yahoo/tagchowder/ParserTest.java:67]: Other UTs were created plainly for coverage or just want to make sure parse() wont throw exception.
I think your use case is more into the functionality part.

Can you create your own handler which extends DefaultHandler (override startElement()) and do something like following code? Then can you make sure you are getting the correct element you want. Just give an idea here.
We need to make sure the functionality you need for AMP is working, making sure parse() function works is just not enough.

`final CustomHandler consumer = new CustomHandler(...);

final Parser p = new Parser();
p.setContentHandler(consumer);
p.setErrorHandler(consumer);
p.setFeature(p.AMP_VALIDATION_FEATURE, true);
p.setDefaultBufferSize(html.length() + ADDITIONAL_BUFF_SIZE);
p.parse(new InputSource(new StringReader(html)));

Assert.assertTrue(consumer.numberOf4email(), 1);
....
`

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java [hunk 9]
```diff
@@ -51,6 +52,69 @@ public void testSampleHtml() throws IOException, SAXException {
...
+    }
+
+    /**
+     * Test AMP feature.
+     *
+     * @throws IOException IOException
+     * @throws SAXException SAXException
+     */
+    @Test
+    public void testAMPFeature() throws IOException, SAXException {
```
Comment (fansu) [inline core/src/test/java/com/yahoo/tagchowder/ParserTest.java:77]: looks good

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java [hunk 10]
```diff
@@ -51,6 +52,69 @@ public void testSampleHtml() throws IOException, SAXException {
...
+
+        //Missing '{' bracket
+        char [] exp4 = {'[', 't', 'e', 'x', 't', '}', ']'};
+        valid = parser.areBracketsBalanced(exp4, 0, exp4.length);
+        Assert.assertEquals(valid, false);
+
+        //no brackets, returns false
+        char [] exp5 = {'t', 'e', 'x', 't'};
+        valid = parser.areBracketsBalanced(exp4, 0, exp5.length);
+        Assert.assertEquals(valid, false);
```
Comment (fansu) [inline core/src/test/java/com/yahoo/tagchowder/ParserTest.java:115]: add for [[text]] and {{text}} as well

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java (hunk 8 above)
Comment (nhant01) [reply core/src/test/java/com/yahoo/tagchowder/ParserTest.java:67]: Where is the UT for the Parser in the past when it was modified?

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java (hunk 10 above)
Comment (nhant01) [reply core/src/test/java/com/yahoo/tagchowder/ParserTest.java:125]: Done

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java (hunk 8 above)
Comment (fansu) [reply core/src/test/java/com/yahoo/tagchowder/ParserTest.java:67]: there was no UT at all for this Parser. When we forked from tagsoup, we need one simpliest UT: ParserTest.

Comment (nhant01) [reply core/src/test/java/com/yahoo/tagchowder/ParserTest.java:67]: This covers in the AMP validator itself.

Comment (fansu) [reply core/src/test/java/com/yahoo/tagchowder/ParserTest.java:67]: yes, I understand that. The new code is added into tagchowder which will live by its own. People may start "copy" code and assume this code part is unbeatable while no UT is there to 100% verify. 

You can port some simple one here, if that is easier.

Comment (nhant01) [reply core/src/test/java/com/yahoo/tagchowder/ParserTest.java:72]: Done

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java [hunk 11]
```diff
@@ -51,6 +53,98 @@ public void testSampleHtml() throws IOException, SAXException {
...
+        List<ParsedHtmlTag> parsedHtmlTagList = customHandler.getParsedHtmlTags();
+        Assert.assertEquals(parsedHtmlTagList.size(), 21);
+
+        ParsedHtmlTag parsedHtmlTag = parsedHtmlTagList.get(0);
+        Assert.assertEquals(parsedHtmlTag.lowerName(), "html");
+        Assert.assertTrue(parsedHtmlTag.hasAttribute("⚡4email"));
+
+        parsedHtmlTag = parsedHtmlTagList.get(7);
+        Assert.assertEquals(parsedHtmlTag.lowerName(), "p");
+        Assert.assertTrue(parsedHtmlTag.hasAttribute("[[text]]"));
```
Comment (fansu) [inline core/src/test/java/com/yahoo/tagchowder/ParserTest.java:83]: Please add these check for all 6 good attributes name in your amphtml.txt

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java [hunk 12]
```diff
@@ -51,6 +53,98 @@ public void testSampleHtml() throws IOException, SAXException {
...
+
+        parsedHtmlTag = parsedHtmlTagList.get(12);
+        Assert.assertEquals(parsedHtmlTag.lowerName(), "p");
+        Assert.assertFalse(parsedHtmlTag.hasAttribute("{[text}"));
```
Comment (fansu) [inline core/src/test/java/com/yahoo/tagchowder/ParserTest.java:87]: Please add these check for all 9 santize attributes name in your amphtml.txt

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java [hunk 13]
```diff
@@ -51,6 +53,98 @@ public void testSampleHtml() throws IOException, SAXException {
...
+     * @throws SAXException SAXException
+     */
+    @Test
+    public void testSampleAMPHtml() throws IOException, SAXException {
+        final String html = getSampleHtml("amphtml.txt");
+        final Parser parser = new Parser();
+        final CustomHandler customHandler = new CustomHandler();
+        parser.setContentHandler(customHandler);
+        parser.setErrorHandler(customHandler);
+        parser.setFeature(Parser.AMP_VALIDATION_FEATURE, true);
```
Comment (fansu) [inline core/src/test/java/com/yahoo/tagchowder/ParserTest.java:69]: have another UT with AMP_VALIDATION_FEATURE=true for html.txt and verify attributes match what you expect (such as html and ParsedHtmlTag)

**Review by fansu — CHANGES_REQUESTED** [review]
Once you finish, please squash all commit into one

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java (hunk 12 above)
Comment (nhant01) [reply core/src/test/java/com/yahoo/tagchowder/ParserTest.java:109]: Done

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java (hunk 11 above)
Comment (nhant01) [reply core/src/test/java/com/yahoo/tagchowder/ParserTest.java:88]: Done

**Comment by fansu** [comment]
Can you also change README.md to reflect latest package version 2.0.18 and also add one line in CHANGES file displaying 

Changes from tagsoup 2.0.18
=========================
- Add AMP validation features to support literal attribute value such as 'html ⚡4email' and [text]

File: core/src/test/java/com/yahoo/tagchowder/ParserTest.java (hunk 13 above)
Comment (nhant01) [reply core/src/test/java/com/yahoo/tagchowder/ParserTest.java:112]: Done

**Comment by nhant01** [comment]
> Once you finish, please squash all commit into one

Done

**Comment by nhant01** [comment]
> Can you also change README.md to reflect latest package version 2.0.18 and also add one line in CHANGES file displaying
> 
> # Changes from tagsoup 2.0.18
> * Add AMP validation features to support literal attribute value such as 'html ⚡4email' and [text]

Done

**Review by fansu — APPROVED** [review]

**Comment by fansu** [comment]
Looks good. @lafa please review.

File: core/src/main/java/com/yahoo/tagchowder/Parser.java [hunk 14]
```diff
@@ -1185,6 +1204,27 @@ private String makeName(final char[] buff, final int offset, final int length) {
...
+                //if the beginning character starts with '[' or '{', ensure the brackets are balanced.
+                //allows [text], [[text]], {text}, {{text}}..etc
+                if (start && (ch == '[' || ch == '{'))  {
+                    if (areBracketsBalanced(buff, offset, length)) {
+                        dst.append(ch);
+                        bracketsAreBalanced = true;
+                    }
+                } else if (start && ch == '⚡') {
+                    //if the beginning character starts with '⚡', ensure that this buffer with offset and length contains "⚡4email"
+                    final StringBuffer str = new StringBuffer(length);
```
Comment (fansu) [inline core/src/main/java/com/yahoo/tagchowder/Parser.java:1217]: Talked to Luis, he does not want any "new" memory allocation, can you fix the code?

File: core/src/main/java/com/yahoo/tagchowder/Parser.java [hunk 15]
```diff
@@ -1194,6 +1234,51 @@ private String makeName(final char[] buff, final int offset, final int length) {
...
+     * @param buff characters
+     * @param offset the buffer offset
+     * @param length the buffer length
+     * @return true if brackets are balanced, otherwise false
+     */
+    public boolean areBracketsBalanced(final char [] buff, final int offset, final int length) {
+        int off = offset;
+        int l = length;
+
+        final Deque<Character> stack = new ArrayDeque<>();
```
Comment (fansu) [inline core/src/main/java/com/yahoo/tagchowder/Parser.java:1248]: by Luis: no "new" memory allocation

**Review by fansu — CHANGES_REQUESTED** [review]
please address luis comments

**Review by fansu — APPROVED** [review]

**Comment by fansu** [comment]
Looks good. based on discussion with luis and nirmal, no need to check for bracket balance

File: core/src/test/java/com/yahoo/tagchowder/AttributesImplTest.java [hunk 16]
```diff
@@ -205,8 +205,9 @@ public void testGetValueWithQName() {
         String value0 = "value0";
         String value1 = "value1";
         attributes.addAttribute(uri, localName, qname, type, value0);
-        attributes.addAttribute(uri, localName, qname, type, value1);
-        Assert.assertEquals(attributes.getValue(qname), value0, "value mismatched");
+        attributes.addAttribute(uri, localName, "1", type, value1);
```
Comment (fansu) [inline core/src/test/java/com/yahoo/tagchowder/AttributesImplTest.java:208]: why this changes? looks like your commit is wrong
